sycl/doc/design/CompileTimeProperties.md

```diff
@@ -532,6 +506,199 @@ optimized.  This puts more pressure on the SPIR-V consumer (e.g. JIT compiler)
 to perform these optimizations.
 
 
+## Properties on a non-global pointer type variable
+Applying properties on a non-global pointer type variable can be done with
+`__bultin_intel_sycl_ptr_annotation` builtin function. Comparing to
+`add_ir_annotations_member`, this builtin function is more flexible as
+it can decorate any pointer variable in the kernel.
+
+
+An example of this is the proposed `annotated_ptr` class.
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = properties<>>
+class annotated_ptr{
+  T *ptr;
+ public:
+  annotated_ptr(T *p) : ptr(p) {}
+};
+
+} // namespace sycl::ext::oneapi
+```
+
+where an example use looks like:
+
+```
+using sycl::ext::oneapi;
+
+void foo(int *p) {
+  annotated_ptr<int, decltype(properties{foo, bar<32>})> aptr(p);
+  *aptr = 1;
+}
+```
+
+To precisely control the location of the annotation information for the targetted
+pointer defereference, We invoke a SYCL builtin in the header.  The builtin
+`__builtin_intel_sycl_ptr_annotation` decorates one of the pointer variables,
```

Comment:
I've just discovered that there is built-in `clang` attribute for inserting custom annotations into LLVM IR: [`__builtin_annotation`](https://github.com/intel/llvm/blob/sycl/clang/test/CodeGen/annotations-builtin.c).

I suggest we explore it more to see if it can be applied for our needs instead of introducing a new custom built-in